[DRAFT] Moved FileSharing.test.ts to use fake adapter

mgamis-msft opened the pull request:
# What
<!--- Describe your changes. -->

# Why
<!--- What problem does this change solve? -->
<!--- Provide a link if you are addressing an open issue. -->

# How Tested
<!--- How did you test your change. What tests have you added. -->

# Process & policy checklist
<!--- Review the list and check the boxes that apply. -->

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.
<!--- If yes, describe the impact. -->

mgamis-msft commented on packages/react-composites/tests/browser/chat/fake-adapter/FileSharing.test.ts:
```diff
@@ -254,48 +225,47 @@ test.describe('Filesharing Message Thread', async () => {
 
-    const page1 = pages[1];
-    await waitForTypingIndicatorHidden(page1);
+  test('contains File Download Card in remote message', async ({ serverUrl, page }) => {
```
I had to create a separate test here with qArg `hasRemoteFileSharingMessage` to get filesharing message from a remote participant to get snapshot 'filesharing-file-download-card-in-received-messages.png'

JamesBurnside commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -27,17 +39,56 @@ export const FakeAdapterApp = (): JSX.Element => {
   ) as ChatAdapterModel;
 
+  // Optional params
+  const useFileSharing = Boolean(params.useFileSharing);
```
avoid `use*` prefix unless this is a react hook
```suggestion
const fileSharingEnabled = Boolean(params.fileSharingEnabled);
```
or if this is doing something more specific e.g. attaching a message to the send box, name this to have that specificity:
```suggestion
const attachFileToInitialSendMessage = Boolean(params.attachFileToInitialSendMessage);
```

JamesBurnside commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -2,18 +2,30 @@
+  createAzureCommunicationChatAdapterFromClient,
+  FileDownloadError,
+  FileDownloadHandler
+} from '../../../../src';
```
nit: Try avoid imports direct from the src's index.ts as it can cause circular import chains

JamesBurnside commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -55,29 +115,50 @@ export const FakeAdapterApp = (): JSX.Element => {
+  model: ChatAdapterModel,
+  hasRemoteFileSharingMessage: boolean
+): Promise<ChatAdapter> {
+  const m = new Model({ asyncDelivery: false });
```
this creates a new Model called `m` but also takes in `model` as a function parameter - what is the difference between these two? (why do we create a new one here?0

JamesBurnside replied:
actually I don't see m used, was it left in from testing things?

JamesBurnside commented on packages/react-composites/tests/browser/chat/app/fake-back-end/Model.ts:
```diff
@@ -48,6 +48,23 @@ export class Model {
+
+  public initializeMessages(threadId: string, messages: BasicChatMessage[]) {
+    messages.forEach((message) => {
+      this.checkedGetThread(message.sender, threadId);
```
Can this be used instead of `this.threadMap[threadId].thread`?

JamesBurnside commented on packages/react-composites/tests/browser/chat/app/fake-back-end/Model.ts:
```diff
@@ -48,6 +48,23 @@ export class Model {
+        ...this.threadMap[threadId].thread.messages,
+        {
+          id: `${this.threadMap[threadId].thread.messages}`,
+          sequenceId: `${this.threadMap[threadId].thread.messages}`,
```
this is a little confusing to read here, can you add a fn to the thread interface that is pushMessages(messages) or similar that handles updating the id, sequenceId and other default values.

JamesBurnside reviewed (approved):
Looks great to me! Make sure to get sign off from someone working on file sharing also

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/fake-back-end/types.ts:
```diff
@@ -42,3 +42,5 @@ export interface NetworkEventModel {
   maxDelayMilliseconds?: number;
 }
+
+export type BasicChatMessage = Pick<ChatMessage, 'sender' | 'senderDisplayName' | 'content' | 'metadata'>;
```
Let's keep the `FakeChatClient` totally isolated from e2e testing concerns.

This `BasicChatMessage` is related to how the tests want to initialize the fake. So move this to the test helpers etc.

See also my comment in `FakeChatClient.ts`

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/fake-back-end/FakeChatClient.ts:
```diff
@@ -85,6 +85,11 @@ export class FakeChatClient implements IChatClient {
     });
   }
 
+  initializeMessages(threadId: string, messages: BasicChatMessage[]): Promise<void> {
```
Let's keep the fake completely isolated from test logic.

The constructor for `FakeChatClient` already accepts a preconstructed `Model` to initialize with. The test should construct the appropriate model and initialize the `FakeChatClient` with the `Model`.

prprabhu-ms replied:
UPDATE:

I see that `Model` does not have to initialize the thread and messages.

You have two options:

1. Add a constructor for `Model` that lets you initialize the data. But use the data types from the headless SDK. The fake uses the headless DTOs. No need to invent another one.
2. Simply use the `ChatClient` that you have created to create a thread, add users and send messages as if you're using the real service.

I recommend (2). It needs not new API in the fake, and is easier to extend (you can get the fake in any state you want by using the public headless API, at least until we want error injection).

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -28,24 +36,61 @@ const params = Object.fromEntries(urlSearchParams.entries());
 
   // Optional params
   const useFrLocale = Boolean(params.useFrLocale);
+  const fileSharingEnabled = Boolean(params.fileSharingEnabled);
```
Instead of adding query args for all these cases, can the test create the appropriate `Model` in some of these cases and pass the serialized `Model` as a single query argument?

The fewer the layers between the test and the eventual model / adapter, the easier it is to understand and add tests.

I know we had some trouble earlier importing the types in the test code. Can we give that another look. We seem to be bending over backwards here to pass all this test data, and it makes working with the e2e tests much harder.

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -28,24 +36,61 @@ const params = Object.fromEntries(urlSearchParams.entries());
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  React.useEffect(() => {
```
Our live test e2e apps are already quite complex and in my opinion not easy to maintain.

Since we're doing new work here, let's learn from the last attempt. I'd recommend splitting off a separate test app for fake file sharing so that the logic there is all related to file sharing only.

Then you can use a single query arg to select that app from index.tsx (or even better, use react-router to pick the right app instead of query arguments).

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -64,31 +118,52 @@ export const FakeAdapterApp = (): JSX.Element => {
+  const localUser = { id: { communicationUserId: nanoid() }, displayName: fakeAdapterArgs.localParticipant };
+  const chatClient = new FakeChatClient(chatClientModel, localUser.id);
+  const participants = remoteParticipants;
+  const splicePosition =
```
What is this dark magic?

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -64,31 +118,52 @@ export const FakeAdapterApp = (): JSX.Element => {
-  const remoteParticipants: ChatParticipant[] = model.remoteParticipants.map((user) => {
+async function createFakeChatAdapter(
+  fakeAdapterArgs: FakeChatAdapterArgs,
+  attachFileToInitialSendMessage: boolean
```
This flag argument is adding a lot of complexity to this function. We should be able to send the first message etc *after* the adapter creation.

prprabhu-ms reviewed (changes requested):
I have substantive comments on this one. Feel free to schedule some time to talk IRL.

mgamis-msft commented:
> I have substantive comments on this one. Feel free to schedule some time to talk IRL.

Hey @prprabhu-ms, I addressed most of the comments and am now able import types in the test code by upgrading playwright. Please take another look.

prprabhu-ms commented on change/@internal-react-composites-4beccd45-629a-4c8a-9905-daf1e36ccee0.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
nit: type should be `none`. Playwright is a `devDependency` that doesn't affect the final package in any way.

prprabhu-ms commented on packages/react-composites/tests/browser/chat/app/FakeAdapterApp.tsx:
```diff
@@ -31,34 +31,82 @@ const params = Object.fromEntries(urlSearchParams.entries());
+          participants
+        }
+      );
+      const adapterInfo = {
```
nitty nit. Reads cleaner: 
```
const newAdapter = await initializeAdapter({ ... });
setAdapter(newAdapter);
```

prprabhu-ms commented on packages/react-composites/tests/browser/chat/fake-adapter/FakeChatAdapterArgs.ts:
```diff
@@ -0,0 +1,18 @@
+
+export type FileUpload = FileMetadata & { uploadComplete?: boolean; error?: string; progress?: number };
+
+export type FakeChatAdapterArgs = {
```
Let's add docstrings to any non-trivial fields here. For example, looking at this I don't understand what `localParticipantPosition` does. When writing a new test, it would be helpful to be able to understand all these args quickly.

prprabhu-ms reviewed (approved):
I see some extra space in a few screenshots: 
![image](https://user-images.githubusercontent.com/82062616/171289174-3ad86081-e512-495f-93d5-c3f730ff9abf.png)

Other than that, LGTM. Worth having Anjul take a look as well.

Kudos for getting these tests converted!
